nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/actions/CSVExportAction.java

```diff
@@ -0,0 +1,309 @@
+import org.nuxeo.runtime.kv.KeyValueStore;
+
+/**
+ * @since 10.3
+ */
+public class CSVExportAction extends AbstractBulkAction {
+
+    public static final String ACTION_NAME = "csvExport";
+
+    public CSVExportAction() {
+        super(ACTION_NAME);
+    }
+
+    @Override
+    public Topology getTopology(Map<String, String> options) {
+        int size = getOptionAsInteger(options, BATCH_SIZE_OPT, DEFAULT_BATCH_SIZE);
+        int timer = getOptionAsInteger(options, BATCH_THRESHOLD_MS_OPT, DEFAULT_BATCH_THRESHOLD_MS);
+        return Topology.builder()
+                       .addComputation(() -> new DocumentCSVProjection(size, timer),
+                               Arrays.asList("i1:" + ACTION_NAME, "o1:" + FileMaterialization.NAME))
+                       .addComputation(FileMaterialization::new,
+                               Arrays.asList("i1:" + FileMaterialization.NAME, "o1:" + SortBlobLines.NAME))
+                       .addComputation(SortBlobLines::new,
+                               Arrays.asList("i1:" + SortBlobLines.NAME, "o1:" + ZipBlob.NAME))
+                       .addComputation(ZipBlob::new,
+                               Arrays.asList("i1:" + ZipBlob.NAME, "o1:" + ExposeBlobInDownload.NAME))
+                       .addComputation(ExposeBlobInDownload::new,
+                               Arrays.asList("i1:" + ExposeBlobInDownload.NAME, "o1:" + COUNTER_ACTION_NAME))
+                       .build();
+    }
+
+    @Override
+    protected Computation createComputation(int size, int timer) {
+        throw new UnsupportedOperationException();
+    }
+
+    public static class DocumentCSVProjection extends AbstractBulkComputation {
+
+        protected StringBuilder output;
+
+        public DocumentCSVProjection(int size, int timer) {
+            super(ACTION_NAME, 1, 1, size, timer);
+        }
+
+        @Override
+        protected void compute(CoreSession session, List<String> ids, Map<String, Serializable> properties) {
+            output = new StringBuilder();
+            for (String id : ids) {
+                // TODO project document into CSV line
+                output.append(id).append(",").append(id).append("\n");
+            }
+        }
+
+        @Override
+        public void createOutput(ComputationContext context) {
+            context.produceRecord("o1", BulkRecords.of(currentCommandId, documentIds.size(), output.toString()));
+        }
+
+    }
+
+    public static class FileMaterialization extends AbstractTransientBlobComputation {
+
+        public static final String NAME = "materlizeAsBlob";
+
+        private Path temp;
+
+        private long count;
+
+        public FileMaterialization() {
+            super(NAME);
+        }
+
+        @Override
+        public void processRecord(ComputationContext context, String documentIdsStreamName, Record record) {
+            String commandId = commandIdFrom(record);
+            appendToFile(commandId, dataFrom(record), countFrom(record));
```

Comment:
It seems that the only place were you need it, if `Record` owns a list of line instead, you can deduce this value directly.
Furthermore, you wouldn't have to worry about `line.separator` of your OS by doing that.

Reply:
Using a single string already joined with \n prevent to join it with , send it then split it and join it again with \n